Packet++/src/IcmpV6Layer.cpp

```diff
@@ -0,0 +1,199 @@
+#define LOG_MODULE PacketLogModuleIcmpV6Layer
+
+#include "IcmpV6Layer.h"
+#include "EndianPortable.h"
+#include "IPv4Layer.h"
+#include "IPv6Layer.h"
+#include "NdpLayer.h"
+#include "PacketUtils.h"
+#include "PayloadLayer.h"
+#include <memory>
+#include <sstream>
+
+// IcmpV6Layer
+
+namespace pcpp
+{
+
+ProtocolType IcmpV6Layer::getIcmpv6Version(uint8_t *data, size_t dataLen)
+{
+	if (dataLen < sizeof(icmpv6hdr))
+		return UnknownProtocol;
+
+	icmpv6hdr *hdr = (icmpv6hdr *)data;
+
+	switch (hdr->type)
+	{
+	case ICMPv6_ECHO_REQUEST:
+		return ICMPv6EchoRequest;
+	case ICMPv6_ECHO_REPLY:
+		return ICMPv6EchoReply;
+	case ICMPv6_NEIGHBOR_SOLICITATION:
+		return NDPNeighborSolicitation;
+	case ICMPv6_NEIGHBOR_ADVERTISEMENT:
+		return NDPNeighborAdvertisement;
+	}
+
+	return UnknownProtocol;
+}
+
+ICMPv6MessageType IcmpV6Layer::getMessageType() const
+{
+	return (ICMPv6MessageType)getIcmpv6Header()->type;
+}
+
+uint8_t IcmpV6Layer::getCode() const
+{
+	return getIcmpv6Header()->code;
+}
+
+uint16_t IcmpV6Layer::getChecksum() const
+{
+	return be16toh(getIcmpv6Header()->checksum);
+}
+
+void IcmpV6Layer::parseNextLayer()
+{
+	size_t headerLen = getHeaderLen();
+	uint8_t *payload = m_Data + getHeaderLen();
+	size_t payloadLen = m_DataLen - getHeaderLen();
+
+	switch (getMessageType())
+	{
+	case ICMPv6_NEIGHBOR_SOLICITATION:
+	case ICMPv6_NEIGHBOR_ADVERTISEMENT:
+		break;
+	default:
+		if (m_DataLen > headerLen)
+			m_NextLayer = new PayloadLayer(m_Data + headerLen, m_DataLen - headerLen, this, m_Packet);

```

Comment:
Isn't ICMPv6 always the last layer? If that's the case let's always consider all of the data as part of the ICMPv6 message even if we don't know how to parse it